Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,176 @@
+        const int argc,
+        const char* argv[],
+        const hostfxr_parameters * parameters);
+```
+
+This new API supercedes the existing `hostfxr_main` and `hostfxr_main_startupinfo`. The function is the main entry point for running applications and dotnet commands. It encapsulates both the `muxer` functionality for running apps via `dotnet app.dll` or `dotnet exec app.dll`, as well as running SDK commands like `dotnet new` and so on.
+
+The proposed behavior is the same as for the existing entry points `hostfxr_main` and `hostfxr_main_startupinfo` with the addition of `parameters` structure. This is meant to:
+* Add the ability to specify additional runtime properties which will be added to the property bag pass to the runtime during its initialization.
+* Make the API extensible without adding new entry point. In the future the parameters structure can be extended with additional members.
+
+The ability to specify additional runtime properties introduces potential collisions with runtime properties initialized by the hosting layer. This ability is not new, it's already possible to specify arbitrary set of additional runtime properties through the `configProperties` section of the `.runtimeconfig.json`. The existing behavior for collisions is somewhat undefined by it effectively means:
+* Properties supplied by the host are added first
+* Any additional properties are added after these - duplicates are retained
+* Runtime initialization will use the last value of any given property it finds for properties it needs to start running managed code
+* Later on during runtime initialization duplicates will cause a failure and the runtime will fail to start with `0x80070057 (E_INVALIDARG)` (internally the failure is due to adding the same key twice to a `Dictionary<K,V>`).
+
+This means that having duplicate properties leads to runtime initialization failure (although the exact set of properties might end up with different failure modes, but it will always fail).
+
+For now, this proposal is not trying to improve this behavior and the additional properties specified through the new `hostfxr` API will behave exactly like those specified via `.runtimeconfig.json`. Since duplicates are not allowed (they cause failures) there's no need to specify precedence (although the logical one would be to take the properties from the API over those in `.runtimeconfig.json`).
+
+## Impact on hosting components
+
+### `hostfxr`
+Aside from the new API `hostfxr_main_with_parameters` the only other improvement is to extend the existing `hostfxr_get_runtime_delegate` to accept additional properties:
+``` C++
+int hostfxr_get_runtime_delegate(
+        const hostfxr_parameters * parameters,
+        hostfxr_delegate_type type,
+        void **delegate);
+```
+
+The functionality of the entry point will remain as is, the only addition is using extensible parameter structure and the ability to pass additional runtime properties.
+
```

Comment:
I don't really understand the rationale here.  Why are the parameters expected to be useful?

Reply:
Having the parameters structure allows us to add to it without introducing new entry points (we simply rev the version of the struct). That's what I meant by saying "extensible". Both `hostfxr` and `hostpolicy` already have cases of APIs which were done one way in V1 and then in V2 we needed new parameters and so introduced effectively api2() (even though it's called differently). I want us to avoid that approach going forward if possible.